File: benchmarks/BunnymarkV1Sprites/cpp/BunnymarkV1Sprites.cpp
#include <stdlib.h>
#include <vector>
#include <tuple>
#include <string>
#include <core/Godot.hpp>
#include <core/GodotGlobal.hpp>
#include <Node2D.hpp>
#include <Texture.hpp>
#include <Sprite.hpp>
#include <ResourceLoader.hpp>
#include <Array.hpp>

using namespace godot;

class BunnymarkV1Sprites : public GodotScript<Node2D> {
        GODOT_CLASS(BunnymarkV1Sprites);
public:
        Vector2 screenSize;
        std::vector<std::tuple<Sprite*, Vector2>> bunnies = {};
        Ref<Texture> TBunny = ResourceLoader::load("res://images/godot_bunny.png");
        float gravity = 500;

        BunnymarkV1Sprites() { srand (time(NULL)); }

        void _ready() {
                owner->set_process(true);
        }

        void _process(const float delta) {
                screenSize = owner->get_viewport_rect().size;
                for (int i = 0; i < bunnies.size(); i++) {
                        std::tuple<Sprite*, Vector2> bunny = bunnies[i];
                        Sprite* bunnySprite = std::get<0>(bunny);
                        Vector2 position = bunnySprite->get_position();
                        Vector2 newPosition = std::get<1>(bunny);

                        position.x += newPosition.x * delta;
                        position.y += newPosition.y * delta;

                        newPosition.y += gravity * delta;

                        if (position.x > screenSize.x)
                        {
                            newPosition.x *= -1;
                            position.x = screenSize.x;
                        }
                        if (position.x < 0)
                        {
                            newPosition.x *= -1;
                            position.x = 0;
                        }

                        if (position.y > screenSize.y)
                        {
                            position.y = screenSize.y;
                            if ((double)rand() / RAND_MAX > 0.5)
                            {
                                newPosition.y = (rand() % 1100 + 50);
                            }
                            else
                            {
                                newPosition.y *= -0.85f;
                            }
                        }

                        if (position.y < 0)
                        {
                            newPosition.y = 0;
                            position.y = 0;
                        }
                        bunnySprite->set_position(position);
                        bunnies[i] = std::make_tuple(bunnySprite, newPosition);
                }
        }

        void add_bunny() {
                Sprite* bunny = new Sprite();
                bunny->set_texture(TBunny.ptr());
                owner->add_child(bunny);
                bunny->set_position(Vector2(screenSize.x / 2.0, screenSize.y / 2.0));
                bunnies.push_back(std::make_tuple(bunny, Vector2(rand()%200+50, rand()%200+50)));
        }

        void remove_bunny() {
                if (bunnies.size() == 0) {
                        return;
                }

                owner->remove_child(std::get<0>(bunnies[bunnies.size() - 1]));
                bunnies.pop_back();
        }

        void finish() {
                Array array;
                array.push_back((uint64_t)bunnies.size());
                owner->emit_signal("benchmark_finished", array);
        }

        static void _register_methods() {
                register_method((char *)"_ready", &BunnymarkV1Sprites::_ready);

                register_method((char *)"_process", &BunnymarkV1Sprites::_process);
                register_method((char *)"add_bunny", &BunnymarkV1Sprites::add_bunny);
                register_method((char *)"remove_bunny", &BunnymarkV1Sprites::remove_bunny);
                register_method((char *)"finish", &BunnymarkV1Sprites::finish);
        }
};

File: benchmarks/BunnymarkV1Sprites/cs/BunnymarkV1Sprites.cs
using Godot;
using System;
using System.Collections.Generic;

public class BunnymarkV1Sprites : Node2D
{
    private struct Pair
    {
        public Sprite Sprite;
        public Vector2 Vector;
    }

    Pair[] bunnies = new Pair[1024];
    int count = 0;
    Vector2 screenSize;

    Texture bunnyTexture = (Texture)GD.Load("res://images/godot_bunny.png");
    Random random = new Random();
    int gravity = 500;

    public override void _Process(float delta)
    {
        screenSize = GetViewportRect().Size;

        for (int i = 0; i < count; i++)
        {
            ref var bunny = ref bunnies[i];
            var position = bunny.Sprite.Position;
            var newPosition = bunny.Vector;

            position.x += newPosition.x * delta;
            position.y += newPosition.y * delta;

            newPosition.y += gravity * delta;

            if (position.x > screenSize.x)
            {
                newPosition.x *= -1;
                position.x = screenSize.x;
            }

            if (position.x < 0)
            {
                newPosition.x *= -1;
                position.x = 0;
            }

            if (position.y > screenSize.y)
            {
                position.y = screenSize.y;
                if (random.NextDouble() > 0.5)
                {
                    newPosition.y = (random.Next() % 1100 + 50);
                }
                else
                {
                    newPosition.y *= -0.85f;
                }
            }

            if (position.y < 0)
            {
                newPosition.y = 0;
                position.y = 0;
            }

            bunny.Sprite.Position = position;
            bunny.Vector = newPosition;
        }
    }

    public void add_bunny()
    {
        var bunny = new Sprite();
        bunny.SetTexture(bunnyTexture);
        AddChild(bunny);
        bunny.Position = new Vector2(screenSize.x / 2, screenSize.y / 2);

        if (count == bunnies.Length)
		{
			Array.Resize(ref bunnies, bunnies.Length * 2);
		}
		bunnies[count] = new Pair() { Sprite = bunny, Vector = new Vector2(random.Next() % 200 + 50, random.Next() % 200 + 50) };
		count++;
    }

    public void remove_bunny()
    {
        if (count == 0) {
			return;
		}

        count--;
        ref var bunny = bunnies[count]
        RemoveChild(bunny.Sprite);
        bunny.Sprite = null;
    }

    public void finish()
    {
        EmitSignal("benchmark_finished", count);
    }
}
